qutebrowser/extensions/loader.py

```diff
@@ -116,6 +124,48 @@ def _walk_normal() -> Iterator[ExtensionInfo]:
         yield ExtensionInfo(name=name)
 
 
+def walk_extensions() -> Iterator[ExtensionInfo]:
+    """Walk external extensions."""
+    ext_dir = os.path.join(standarddir.data(), 'extensions')
+    if not os.path.exists(ext_dir):
+        try:
+            os.mkdir(ext_dir)
+        except FileExistsError:
+            pass
+        except OSError:
+            log.extensions.exception("Failed to create extensions directory")
+        return
+
+    if os.listdir(ext_dir):
+        log.extensions.warning(
+            "Loading EXTENSIONS in {}: this in an "
+            "EXPERIMENTAL FEATURE.".format(ext_dir)
+        )
+    else:
+        return
+
+    prefix = 'qutebrowser.extensions.third_party.'
+    for finder, name, _ispkg in pkgutil.walk_packages(
+            path=[ext_dir],
+            prefix=prefix,
+            onerror=_on_walk_error):
+        if name not in sys.modules:
+            try:
+                # Import the module with the finder so that it is in
+                # sys.modules ready for _load_component.
+                spec = finder.find_spec(name, None)
+                if not spec or not spec.loader:
+                    raise ImportError(f"pkgutil couldn't find loader for {name}")
+                spec.loader.load_module(name)
```

Comment:
TODO: 
> Deprecated since version 3.4: The recommended API for loading a module is exec_module() (and create_module()). Loaders should implement it instead of load_module(). The import machinery takes care of all the other responsibilities of load_module() when exec_module() is implemented.

Reply:
I've pushed a commit that should work with python 3.12. Although I'm not convinced it's the right way to do dynamic imports this. Mainly because I have to add stuff to sys.modules myself which feels wrong. Although there are other examples on the internet doing the [same thing](https://betterstack.com/community/questions/how-to-import-python-module-dynamically/#checking-if-a-module-can-be-imported). See the commit message for more context and ideas.

I forgot to put it in the commit message but this was the stack trace before I was adding it to sys.modules myself:

```
  File "/home/user/projects/qutebrowser/qutebrowser/app.py", line 141, in init
    loader.load_extensions()
  File "/home/user/projects/qutebrowser/qutebrowser/extensions/loader.py", line 88, in load_extensions
    _load_component(info, skip_hooks=skip_hooks)
  File "/home/user/projects/qutebrowser/qutebrowser/extensions/loader.py", line 198, in _load_component
    info.spec.loader.exec_module(mod)
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/tmp/qute-basdir-extensions/data/extensions/jmatrix.py", line 6, in <module>
    @hook.init()
     ^^^^^^^^^^^
  File "/home/user/projects/qutebrowser/qutebrowser/api/hook.py", line 37, in __call__
    info = _add_module_info(func)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/projects/qutebrowser/qutebrowser/api/hook.py", line 18, in _add_module_info
    module = importlib.import_module(func.__module__)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1126, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1140, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'qutebrowser.extensions.third_party'
```